method/stardust/stardust.r

```diff
@@ -181,93 +198,47 @@ if (technology %in% c("ST", "Visium")){
-      }
-      lb <- res
-    } else if (n_clust < n_clust_target) {
-      while (n_clust < n_clust_target) {
-        lb <- res 
-        res <- res*resolution_update
-        results <- do_clustering(countMatrix, res = res, ...)
-        n_clust <- length(unique(results@active.ident))
-      }
-      rb <- res
-    }
-    if (n_clust == n_clust_target) {lb = rb = res }
-  }
+# Clustering
+options(future.globals.maxSize = 100 * 1000 * 1024^2)
 
-  i <- 0
-  while ((rb - lb > tolerance || lb == rb) && i < num_rs) {
-    mid <- sqrt(lb * rb)
-    message("Resolution: ", mid)
-    set.seed(seed)
-    results <- do_clustering(countMatrix, res = mid, ...)
-    cluster_ids <- results@active.ident
-    # Adjust cluster_ids extraction per method
-    n_clust <- length(unique(cluster_ids))
-    message("Cluster: ", n_clust)
-    if (n_clust == n_clust_target || lb == rb) break
-    if (n_clust > n_clust_target) {
-      rb <- mid
-    } else {
-      lb <- mid
-    }
-    i <- i + 1
-  }
+countMatrix <- countMatrix[,sort(colnames(countMatrix))]
+d <- dim(spotPositions)[2]
+spotPositions <- spotPositions[,(d-1):d]
+spotPositions <- spotPositions[sort(rownames(spotPositions)),]
 
-  # Warning if target not met
-  if (n_clust != n_clust_target) {
-    warning(sprintf("Warning: n_clust = %d not found in binary search, return best approximation with res = %f and n_clust = %d. (rb = %f, lb = %f, i = %d)", n_clust_target, mid, n_clust, rb, lb, i))
-  }
-  return(results)
+seuratObject <- Seurat::CreateSeuratObject(countMatrix)
+seuratObject <- suppressWarnings(Seurat::SCTransform(seuratObject, assay = "RNA", verbose = FALSE))
+seuratObject <- Seurat::RunPCA(seuratObject, assay = "SCT", verbose = FALSE)
+if(pcaDimensions <= 2){
+  pcaDimensions = 2
 }
-
-# Clustering
+m <- seuratObject@reductions[["pca"]]@cell.embeddings[,1:pcaDimensions]
+distPCA = dist(m,method="minkowski",p=2)
+distCoord <- dist(spotPositions,method="minkowski",p=2)
+distCoord <- distCoord*(max(distPCA)/max(distCoord))
 
 if (method == "weight"){
-    do_clustering <- weightStardust
-    formals(do_clustering)$spaceWeight <- config$weight
+  spaceWeight <- config$weight
+  distCoord <- distCoord*((max(distPCA)*as.double(spaceWeight))/(max(distCoord)))
 } else {
-    do_clustering <- autoStardust
+  expr_norm <- (distPCA - min(distPCA)) / (max(distPCA) - min(distPCA))
+  distCoord <- (distCoord)*(as.double(as.vector(expr_norm)))
 }
 
-options(future.globals.maxSize = 100 * 1000 * 1024^2)
+finalDistance <- as.matrix(distPCA + distCoord)
+neighbors <- suppressWarnings(Seurat::FindNeighbors(finalDistance))
+neighbors <- list(neighbors_nn=neighbors$nn,neighbors_snn=neighbors$snn)
+seuratObject@graphs <- neighbors
+
+do_clustering <- Seurat::FindClusters
 
 output <- binary_search(
-    countMatrix, n_clust_target = n_clusters, 
+    countMatrix, 
+    n_clust_target = n_clusters, 
     do_clustering = do_clustering, 
     # stardust specific
```

Comment:
Forgot to mention that the `binary_search` function also requires an `extract_nclust` parameter, which is a function to count the number of clusters. In this case, it could be:
`extract_nclust = function(results) length(unique(results@active.ident))`